catalog/app/containers/UriResolver/UriResolver.js

```diff
@@ -0,0 +1,139 @@
+import * as React from 'react'
+import { useHistory, Link, Redirect } from 'react-router-dom'
+import * as M from '@material-ui/core'
+
+import Layout from 'components/Layout'
+import * as NamedRoutes from 'utils/NamedRoutes'
+import * as PackageUri from 'utils/PackageUri'
+
+const useStyles = M.makeStyles((t) => ({
+  form: {
+    display: 'flex',
+    marginTop: t.spacing(3),
+  },
+  btn: {
+    marginLeft: t.spacing(2),
+  },
+  linkBox: {
+    marginTop: t.spacing(4),
+    marginLeft: 'auto',
+    marginRight: 'auto',
+  },
+  link: {
+    display: 'block',
+    padding: t.spacing(1.5, 2),
+    '&:hover': {
+      background: t.palette.action.hover,
+    },
+  },
+  field: {
+    ...t.typography.body1,
+    display: 'block',
+    overflow: 'hidden',
+    textOverflow: 'ellipsis',
+    whiteSpace: 'nowrap',
+  },
+  name: {
+    fontWeight: t.typography.fontWeightMedium,
+  },
+}))
+
+export default function UriResolver({ match }) {
+  const { urls } = NamedRoutes.use()
+  const history = useHistory()
+  const classes = useStyles()
+
+  const uri = decodeURIComponent(match.params.uri || '')
+  const [parsed, error] = React.useMemo(() => {
+    try {
+      return [uri ? PackageUri.parse(uri) : null]
+    } catch (e) {
+      return [null, e]
+    }
+  }, [uri])
+
+  const [value, setValue] = React.useState(uri)
+
+  const handleChange = React.useCallback(
+    (e) => {
+      setValue(e.target.value)
+    },
+    [setValue],
+  )
+
+  const handleSubmit = React.useCallback(
+    (e) => {
+      e.preventDefault()
+      if (value !== uri) history.push(urls.uriResolver(value))
+    },
+    [value, uri, history, urls],
+  )
+
+  const to =
+    parsed &&
+    urls.bucketPackageTree(
+      parsed.bucket,
+      parsed.name,
+      parsed.hash || parsed.tag,
+      parsed.path,
+    ) + NamedRoutes.mkSearch({ resolvedFrom: uri })
+
+  // automatically redirect if URI is pre-filled with a correct value
+  const [redirect] = React.useState(!!to)
```

Comment:
It's weird that there is no setter.
`const redirect = useMemo(() => !!to, [])` doesn't fit?

Reply:
i need this to be executed only once, on construction. `useMemo` does not give us that semantic guarantee and should be used only for optimization